beacon-chain/sync/initial-sync/blocks_queue.go

```diff
@@ -0,0 +1,413 @@
+		quit:                 make(chan struct{}),
+	}
+}
+
+// start boots up the queue processing.
+func (q *blocksQueue) start() error {
+	select {
+	case <-q.ctx.Done():
+		return errQueueCtxIsDone
+	default:
+		go q.loop()
+		return nil
+	}
+}
+
+// stop terminates all queue operations.
+func (q *blocksQueue) stop() error {
+	q.cancel()
+	select {
+	case <-q.quit:
+		return nil
+	case <-time.After(queueStopCallTimeout):
+		return errQueueTakesTooLongToStop
+	}
+}
+
+// loop is a main queue loop.
+func (q *blocksQueue) loop() {
+	defer close(q.quit)
+
+	// Wait for all goroutines to wrap up (forced by cancelled context), and do a cleanup.
+	wg := &sync.WaitGroup{}
+	defer func() {
+		wg.Wait()
+		q.blocksFetcher.stop()
+		close(q.fetchedBlocks)
+	}()
+
+	if err := q.blocksFetcher.start(); err != nil {
+		log.WithError(err).Debug("Can not start blocks provider")
+	}
+
+	// Reads from semaphore channel, thus allowing next goroutine to grab it and schedule next request.
+	releaseTicket := func() {
+		select {
+		case <-q.ctx.Done():
+		case <-q.pendingFetchRequests:
+		}
+	}
+
+	for {
+		if q.headFetcher.HeadSlot() >= q.highestExpectedSlot {
+			log.Debug("Highest expected slot reached")
+			q.cancel()
+		}
+
+		select {
+		case <-q.ctx.Done():
+			log.Debug("Context closed, exiting goroutine (blocks queue)")
+			return
+		case q.pendingFetchRequests <- struct{}{}:
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+
+				// Schedule request.
+				if err := q.scheduleFetchRequests(q.ctx); err != nil {
+					q.state.scheduler.incrementCounter(failedBlock, blockBatchSize)
+					releaseTicket()
+				}
+			}()
+		case response, ok := <-q.blocksFetcher.requestResponses():
+			if !ok {
+				log.Debug("Fetcher closed output channel")
+				q.cancel()
+				return
+			}
+
+			// Release semaphore ticket.
+			go releaseTicket()
+
+			// Process incoming response into blocks.
+			wg.Add(1)
+			go func() {
+				defer func() {
+					select {
+					case <-q.ctx.Done():
+					case q.pendingFetchedBlocks <- struct{}{}: // notify sender of data availability
+					}
+					wg.Done()
+				}()
+
+				skippedBlocks, err := q.parseFetchResponse(q.ctx, response)
+				if err != nil {
+					q.state.scheduler.incrementCounter(failedBlock, response.count)
+					return
+				}
+				q.state.scheduler.incrementCounter(skippedBlock, skippedBlocks)
+			}()
+		case <-q.pendingFetchedBlocks:
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				if err := q.sendFetchedBlocks(q.ctx); err != nil {
+					log.WithError(err).Debug("Error sending received blocks")
+				}
+			}()
+		}
+	}
+}
+
+// scheduleFetchRequests enqueues block fetch requests to block fetcher.
+func (q *blocksQueue) scheduleFetchRequests(ctx context.Context) error {
+	q.state.scheduler.Lock()
+	defer q.state.scheduler.Unlock()
+
+	if ctx.Err() != nil {
+		return ctx.Err()
+	}
+
+	s := q.state.scheduler
+	blocks := q.state.scheduler.requestedBlocks
+
+	func() {
```

Comment:
why make this a function it is only ever called once

Reply:
I needed that logic isolated (so that I can return under one condition or accrue several ifs/effects on other situations). Making this function not inline, makes it harder to manage mutex locks.